pkg/apis/pipeline/v1/result_validation.go

```diff
@@ -33,21 +33,23 @@ func (tr TaskResult) Validate(ctx context.Context) (errs *apis.FieldError) {
 	if !resultNameFormatRegex.MatchString(tr.Name) {
 		return apis.ErrInvalidKeyName(tr.Name, "name", fmt.Sprintf("Name must consist of alphanumeric characters, '-', '_', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my-name',  or 'my_name', regex used for validation is '%s')", ResultNameFormat))
 	}
-	// Array and Object are alpha features
-	if tr.Type == ResultsTypeArray || tr.Type == ResultsTypeObject {
+
+	switch {
+	// Object are alpha features
+	case tr.Type == ResultsTypeObject:
 		errs := validateObjectResult(tr)
 		errs = errs.Also(version.ValidateEnabledAPIFields(ctx, "results type", config.AlphaAPIFields))
 		return errs
-	}
-
+	// Array results is a beta feature - check if the feature flag is set to "beta" or "alpha"
+	case tr.Type == ResultsTypeArray:
+		errs = errs.Also(version.ValidateEnabledAPIFields(ctx, "results type", config.BetaAPIFields))
+		return errs
 	// Resources created before the result. Type was introduced may not have Type set
 	// and should be considered valid
-	if tr.Type == "" {
+	case tr.Type == "":
 		return nil
-	}
-
-	// By default the result type is string
-	if tr.Type != ResultsTypeString {
+	// By default, the result type is string
+	case tr.Type != ResultsTypeString:
```

Comment:
NIT: I find it a bit brittle to rely on the sum of case condition to provide full coverage of all cases, I think it would be preferable to do something like:

```go
	default:
            if tr.Type != ResultsTypeString:
               return apis.ErrInvalidValue(tr.Type, "type", "type must be string")
```

Reply:
thank you @afrittoli I am not able to follow the concern here, sorry 🙃 😕

Reply:
the `default` is always guaranteed to execute, not sure if it make sense to introduce default with a `condition` 🤔 v/s its better to be explicit with the `case` statement ...

Reply:
The `default` is not always guaranteed to execute, it won't be executed if any of the options before matches. 
But it guarantees execution if the value was not caught by any of the `case` statements. 
My proposed alternative is not very different indeed, I think a better one would be to explicitly succeed on "string" and fail with a different error message in the default case.
Not blocking for this PR anyways, I can make a small update once this is merged.